packages/playwright-ct-angular/index.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const { test, expect, devices, defineConfig: originalDefineConfig } = require('@playwright/experimental-ct-core');
+const path = require('path');
+
+process.env['NODE_ENV'] = 'test';
+
+function plugin() {
+  // Only fetch upon request to avoid resolution in workers.
+  const { createPlugin } = require('@playwright/experimental-ct-core/lib/vitePlugin');
+  return createPlugin(
+      path.join(__dirname, 'registerSource.mjs'),
+      () => import('@analogjs/vite-plugin-angular').then(plugin => {
```

Comment:
IMO, the most important conversation is about `@analogjs/vite-plugin-angular`

## 1. should `@analogjs/vite-plugin-angular` be optional?
Adding a vite plugin could be an overhead in some cases. I'm not here to advocate for Angular Single-File-Component (inline template / styles), but there could be issues (cf. https://github.com/analogjs/analog/issues/410)


## 2. Users should probably be able to customize `@analogjs/vite-plugin-angular` 
In most SFC cases, we don't need transforms and jit turns out to be faster so skipping the plugin is an interesting option.
I also noticed some issues when the plugin doesn't grab the right tsconfig etc... so we need to be able to pass the tsconfig to the plugin.

Shouldn't we allow users to pass options to the plugin?

## 3. `@analogjs/vite-plugin-angular` should be an opt-out
In most cases, we need the plugin so it should be enabled by default while we could still disable it if needed.

## 4. `@analogjs/vite-plugin-angular` should not be a dependency

Now the main problem with the current approach is that the current package is highly coupled to `@analogjs/vite-plugin-angular` which is coupled to Angular's major version and I guess that it could even break on Angular minors because Angular's eslint plugins are not public API.

Suppose `@analogjs/vite-plugin-angular` 0.3.0 is only compatible with Angular >=17, while v0.2.0 is only compatible with <=16. Bumping is going to be a hard decision, especially since playwright CT packages do not follow semver. We might not want to have different plugins for different Angular versions 🙈.

a. Maybe `@analogjs/vite-plugin-angular` should be a peer dependency.
b. Maybe https://github.com/microsoft/create-playwright should add the dependency and set up `playwright.config.ts` with `@analogjs/vite-plugin-angular` which is passed as a parameter to `defineConfig`.
c. Or my favorite option, always use the latest version of `@analogjs/vite-plugin-angular` as a dependency but let users control the plugin and not just the options so that they can provide an older version or a future unstable version if they live on t he bleeding edge or another plugin if they have one.

cc. @brandonroberts

Reply:
Thanks @yjaaidi. The thing is, it won't work without a plugin. You could use JIT mode, but unless you have the ability to use inline templates, inline styles, and inject everywhere including library code, things will break. 

Even if there was first-party support by Angular, it would still need to be done through a plugin.

The plugin is what connects Angular compilation to work within Vite, which is what Playwright uses for component testing. The plugin itself its built to support Angular >=16 and its intended to be kept that way and not be tied to specific versions of Angular.

However you want to add the plugin is fine, but its not optional in this implementation.

Reply:
@yjaaidi @analogjs/vite-plugin-angular would be a tough dependency to add:

- it comes from a new publisher
- it does not have a ton of dependencies or downloads yet
- it does not seem to have a lot of code

Unless we anticipate a ton of maintenance work around it, I would consider implementing a plugin tailored around the Playwright needs instead. That would help with upstream fixes, versioning, etc. This is not a strict requirement while we are experimental, but that would become a GA blocker.

@brandonroberts I don't want it to come across as not appreciation of your hard work. I think your plugin is great, not seeing official support for Angular for Vite is quite unfortunate. It is just that the acceptance criteria for the npm modules Playwright can depend on is quite strict - we need to anticipate all kinds of risks and maintenance costs.

Reply:
@pavelfeldman no offense taken. I think your concerns are definitely valid and official support for Vite with Angular through a plugin would be ideal. I'll leave it up to you all on how you want to proceed from here.

Reply:
Angular v17 comes with the vite dev-server by default, so it's quite stable. What do you think about directly reusing the [internal Vite dev-server setup function from the CLI](https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/builders/dev-server/vite-server.ts#L349) for playwright? Note that Angular's CLI programmatic APIs are internal and subject to change without following SemVer, but this way, we don't add any new dependencies, and we stick to the framework's internals.

Reply:
@edbzn I think that this would require lots of changes to Playwright and also the Angular vite server function as it would have to allow us customizing `entries` and probably a couple of other options.

Reply:
> The plugin itself its built to support Angular >=16 and its intended to be kept that way and not be tied to specific versions of Angular.

Oh! That's good news!

Reply:
@brandonroberts don't get me wrong. Currently `@analogjs/vite-plugin-angular` is undeniably the best option for integrating Angular with other tools.

## Concerning JIT mode
Concerning JIT mode, the only prerequisites are inline templates / inline styles / inject everywhere _(but AFAIK not in libraries. Libraries are built with partial mode which generates the `ɵfac` factory using `ɵɵinject`)_

Should this be the default setup? I don't think so. We can't make these prerequisites to Playwright Component Testing, but it should still be an option.

## `@analogjs/vite-plugin-angular` as a dep or not?

I think that maintaining an additional plugin is not an option. It's probably not much code but I know that it requires lots of research and maintenance to hook into `@angular-devkit/build-angular` internals. Cf. https://github.com/analogjs/analog/blob/main/packages/vite-plugin-angular/src/lib/utils/devkit.ts

The Angular team might have some future plans to provide a plugin but there is nothing official yet.
Let's hope that this happens before CT GA 😊.

As Brandon confirms that "The plugin itself its built to support Angular >=16 and its intended to be kept that way and not be tied to specific versions of Angular", then the versioning problem I mentioned before disappears. So the plugin could be added as a dependency.

## How to customize `@analogjs/vite-plugin-angular` options

Finally, in order to customize the plugin's options, I think that developers should pass the plugin as an option to `defineConfig` instead of passing the plugin's options.
Exposing the plugin's options in the `defineConfig` would highly couple this package to the plugin.

## TL;DR
My general conclusions:
1. `@analogjs/vite-plugin-angular` can be added as a dep meanwhile Angular exposes an official plugin
2. `defineConfig` should allow us to pass `plugins` (e.g. this way developers can use their custom plugins, or override `@analogjs/vite-plugin-angular`)

What do you think?

Reply:
I just noticed that Playwright CT is using Vite's `mergeConfig` to merge `ctViteConfig` with the config embedded in  `@playwright/experiment-ct-angular`.
This means that there is currently no way of choosing a different Vite plugin.

A. Should we move this to "user-land" and let `create-playwright` add `@analogjs/vite-plugin-angular` to `package.json` and set up `playwright.config.ts`'s `ctViteConfig` accordingly?
B. Should `defineConfig()` provide an additional opt-out like `ctDisableBuiltInPlugins`?
C. Should `defineConfig()` detect when `ctViteConfig.plugins` is set but that is not aligned with other plugins behavior and can be hard to maintain?

cc. @mxschmitt @pavelfeldman @sand4rt

Reply:
I believe that @yjaaidi has made the optimal choice _(to move `@analogjs/vite-plugin-angular` to `create-playwright`)_. In this scenario, @brandonroberts's Vite plugin is still used _(as I don't see much value without Vite plugin)_. Hopefully, it will be easier to gain approval from @pavelfeldman since the library is not in the main repository. See also the other pros and cons mentioned here: https://github.com/microsoft/playwright/pull/27783#issuecomment-1821277873.

It would be great if there were an official Vite plugin for Angular. @mgechev, do you have any information on whether this is gonna happen?